File: FixedReceiptData/FixedReceiptData/International/AmountDataFixed.cpp
#include <iostream>
#include <sstream>
#include <set>
#include <regex>

#include "../Utils.h"
#include "AmountDataFixed.h"

namespace International
{
	AmountDataFixed::AmountDataFixed(const std::string &InValidatedPath, const std::string &InResultPath)
		: DataFixed(InValidatedPath, InResultPath) {

	}

	AmountDataFixed::~AmountDataFixed() {

	}

	void AmountDataFixed::BeforeFixed() {
		std::cout << "Start Fixed Amount Data=================>" << std::endl;
		std::cout << std::endl;
	}

	void AmountDataFixed::FixedData(std::shared_ptr<rapidjson::Document> &InValidatedDocument,
		std::shared_ptr<rapidjson::Document> InResultDocument) {

		std::string validated_subtotal, validated_tip, validated_tiprate, validated_total;
		std::string result_subtotal, result_tip, result_tiprate, result_total;

		if (!ParseData(InValidatedDocument, validated_subtotal, validated_tip, validated_tiprate, validated_total)
			|| !ParseData(InResultDocument, result_subtotal, result_tip, result_tiprate, result_total)
			|| !CheckData(validated_subtotal)
			|| !CheckData(validated_tip)
			|| !CheckData(validated_total)) {
			std::cout << " Data Error" << std::endl;
			return;
		}

		if (!Equal(validated_subtotal, result_subtotal) || !Equal(validated_tip, result_tip) || !Equal(validated_total, result_total)) {
			ErrorCount++;
		}
		else {
			std::cout << "Validated Equal To Result" << std::endl;
			return;
		}

		std::cout << result_subtotal << ", " << result_tip << ", " << result_total << std::endl;
		std::cout << (result_tiprate.length() == 0 ? "(Empty)" : result_tiprate) << std::endl;
		std::cout << " Fixed To " << std::endl;

		NormalizeData(result_subtotal);
		NormalizeData(result_tip);
		NormalizeData(result_total);

		std::string tiprate_tip, tiprate_total;
		FixedTipRate(result_tiprate, tiprate_tip, tiprate_total);
		FixedAmountData(result_subtotal, result_tip, result_total, tiprate_tip, tiprate_total);

		if (!CheckData(result_subtotal))
			FixedAmountData(result_subtotal);
		if (!CheckData(result_tip))
			FixedAmountData(result_tip);
		if (!CheckData(result_total))
			FixedAmountData(result_total);

		std::cout << result_subtotal << ", " << result_tip << ", " << result_total << std::endl;
		std::cout << (result_tiprate.length() == 0 ? "(Empty)" : result_tiprate) << std::endl;

		if (Equal(validated_subtotal, result_subtotal) && Equal(validated_tip, result_tip) && Equal(validated_total, result_total)) {
			FixedCount++;
			std::cout << "Fixed Success!" << std::endl;
		}
		else {
			std::cout << "Validated " << validated_subtotal << ", " << validated_tip << ", " << validated_total << std::endl;
			std::cout << "Fixed Falied!" << std::endl;
		}
	}

	void AmountDataFixed::AfterFixed() {
		std::cout << "Error Count " << ErrorCount << ", Fixed Count " << FixedCount << ", Fixed Error Count " << FixedErrorCount << std::endl;

		std::cout << std::endl;
		std::cout << "<=================End Fixed Amount Data" << std::endl;
	}

	bool AmountDataFixed::ParseData(const std::shared_ptr<rapidjson::Document> &jsonDocument, std::string &subtotal, std::string &tip, std::string &tip_rate, std::string &total) {
		if (!jsonDocument || !jsonDocument->IsObject() || !jsonDocument->HasMember("regions"))
			return false;

		rapidjson::Value &regions = jsonDocument->operator[]("regions");

		bool has_payment = false, has_tip = false, has_tiprate = false, has_total = false;
		for (auto &region : regions.GetArray()) {
			if (!region.HasMember("cls")
				|| !region.HasMember("result")) {
				continue;
			}

			switch (region["cls"].GetInt())
			{
			case 3: {
				subtotal = region["result"].GetArray()[0].GetString();
				has_payment = subtotal.size();
				break;
			}
			case 5: {
				tip = region["result"].GetArray()[0].GetString();
				has_tip = tip.size();
				break;
			}
			case 6: {
				total = region["result"].GetArray()[0].GetString();
				has_total = total.size();
				break;
			}
			case 11: {
				tip_rate = region["result"].GetArray()[0].GetString();
				has_tiprate = tip_rate.size();
				break;
			}
			default:
				break;
			}
		}

		return has_payment && (has_tip || has_total);
	}

	bool AmountDataFixed::CheckData(const std::string &data)
	{
		if (!CheckNumber(data)) {
			return false;
		}

		size_t dotIndex = data.rfind('.');
		if (dotIndex != std::string::npos && (data.length() - data.rfind('.')) == 3)
			return true;

		return false;
	}

	bool AmountDataFixed::CheckNumber(const std::string &data) {
		if (data.length() == 0) return false;

		static std::set<char> numberPatterns = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.' };
		for (auto ch : data) {
			if (numberPatterns.count(ch) == 0) {
				return false;
			}
		}

		return true;
	}

	void AmountDataFixed::FixedTipRate(std::string &tiprate, std::string &tip, std::string &total) {
		if (tiprate.size() == 0) return;

		std::vector<std::string> tokens;
		std::regex re("%|\\$|,|:|-|\\s+");
		std::sregex_token_iterator token_iterator(tiprate.begin(), tiprate.end(), re, -1);
		std::sregex_token_iterator end;
		while (token_iterator != end) {
			if (CheckNumber(*token_iterator)) {
				tokens.push_back(*token_iterator);
			}
			token_iterator++;
		}

		if (tokens.size() == 3) {
			std::ostringstream stringStream;
			int rate = atoi(tokens[0].c_str());
			if (rate > 100) {
				tokens[0] = std::to_string(rate % 100);
			}
			stringStream << tokens[0] << "% tip = $" << tokens[1] << ", Total = $" << tokens[2];
			tiprate = stringStream.str();
			tip = tokens[1];
			total = tokens[2];
		}
	}

	void AmountDataFixed::FixedAmountData(std::string &data) {
		if (data.length() == 0) return;

		size_t dotIndex = data.rfind('.');
		if (dotIndex != std::string::npos) {
			if ((data.length() - dotIndex) > 2) {
				data = data.substr(0, dotIndex + 3);
			}
			else if ((data.length() - dotIndex) == 2) {
				data.append("0");
			}
			else if ((data.length() - dotIndex) == 1) {
				data.append("00");
			}
		}
		else {
			data.append(".00");
		}
	}

	void AmountDataFixed::FixedAmountData(std::string &subtotal, std::string &tip, std::string &total, const std::string &tiprate_tip, const std::string &tiprate_total) {
		if (CheckNumber(subtotal) && CheckNumber(tip) && CheckNumber(total)) {
			double d_subtotal = std::stod(subtotal), d_tip = std::stod(tip), d_total = std::stod(total);
			if (Equal(d_total, d_subtotal + d_tip)) {
				return;
			}
		}

		if (CheckNumber(tiprate_tip) && CheckNumber(tiprate_total)) {
			double d_tip = std::stod(tiprate_tip), d_total = std::stod(tiprate_total);
			if (CheckNumber(subtotal)) {
				double d_subtotal = std::stod(subtotal), d_tip = std::stod(tiprate_tip), d_total = std::stod(tiprate_total);
				if (Equal(d_total, d_subtotal + d_tip)) {
					total = tiprate_total;
					tip = tiprate_tip;
					return;
				}
			}
			else {
				double d_subtotal = d_total - d_tip;
				std::string fixed = DoubleToString(d_subtotal);
				if (subtotal.length() == 0 || !CheckNumber(subtotal) || Similarity(fixed, subtotal) > 0.7) {
					subtotal = fixed;
					total = tiprate_total;
					tip = tiprate_tip;
					return;
				}
			}
		}


		if (CheckData(subtotal) && FixedDataBySubTotal(subtotal, tip, total)) {
			return;
		}

		if (CheckData(total) && FixedDataByTotal(subtotal, tip, total)) {
			return;
		}

		return;
	}

	bool AmountDataFixed::FixedDataBySubTotal(std::string &subtotal, std::string &tip, std::string &total) {
		double d_subtotal = std::stod(subtotal);

		if (CheckNumber(tip)) {
			double d_tip = std::stod(tip);
			double d_total = d_subtotal + d_tip;
			std::string fixed = DoubleToString(d_total);
			if (total.length() == 0 || !CheckNumber(total) ||Similarity(fixed, total) > 0.7) {
				total = fixed;
				if (!CheckData(tip))
					FixedAmountData(tip);
				return true;
			}
		}

		if (CheckNumber(total)) {
			double d_total = std::stod(total);
			double d_tip = d_total - d_subtotal;
			std::string fixed = DoubleToString(d_tip);
			if (tip.length() == 0 || !CheckNumber(tip) || Similarity(fixed, tip) > 0.7) {
				tip = fixed;
				if (!CheckData(total))
					FixedAmountData(total);
				return true;
			}
		}

		return false;
	}

	bool AmountDataFixed::FixedDataByTotal(std::string &subtotal, std::string &tip, std::string &total) {
		double d_total = std::stod(total);

		if (CheckNumber(tip)) {
			double d_tip = std::stod(tip);
			double d_subtotal = d_total - d_tip;
			std::string fixed = DoubleToString(d_subtotal);
			if (subtotal.length() == 0 || !CheckNumber(subtotal) || Similarity(fixed, subtotal) > 0.7) {
				subtotal = fixed;
				if (!CheckData(tip)) {
					FixedAmountData(tip);
				}
				return true;
			}
		}

		if (CheckNumber(subtotal)) {
			double d_subtotal = std::stod(subtotal);
			double d_tip = d_total - d_subtotal;
			std::string fixed = DoubleToString(d_tip);
			if (tip.length() == 0 || !CheckNumber(tip) || Similarity(fixed, tip) > 0.7) {
				tip = fixed;
				if (!CheckData(subtotal)) {
					FixedAmountData(subtotal);
				}
				return true;
			}
		}

		return false;
	}

	std::string AmountDataFixed::DoubleToString(double invalue)
	{
		std::string value = std::to_string(int(invalue * 100));
		if (value.length() >= 3) {
			value.insert(value.end() - 2, '.');
		}
		else {
			value = std::to_string(invalue);
			size_t dotIndex = value.find('.');
			if (dotIndex != std::string::npos) {
				return value.substr(0, dotIndex + 3);
			}
		}
		return value;
	}

	double AmountDataFixed::Round(double value)
	{
		int ivalue = int(value * 1000);
		if (ivalue % 10 > 4)
		{
			return value + 0.01;
		}

		return value;
	}

	void AmountDataFixed::NormalizeData(std::string &data) {
		if (data.length() == 0) return;

		static std::vector<std::string> patterns = {" ", "$", "\\"};
		for (const auto &pattern : patterns) {
			StringReplace(data, pattern, "");
		}
		/*std::vector<std::string> invalidch;
		static std::set<char> numberPatterns = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.' };
		for (int index = 0; index < data.size(); ) {
			if (numberPatterns.count(data[index]) == 0) {
				data.erase(data.begin() + index);
			}
			else {
				index++;
			}
		}*/
	}
}

File: DataFixed/Python/Receipt/International/AmountDataFixed.py
from decimal import Decimal
import numpy as np
import re
import logging

from DataFixed import DataFixed
from ConfidenceLevel import ConfidenceLevel
from utils import StrUtil

class AmountDataFixed(DataFixed):
    """description of class"""
    def __init__(self):
        DataFixed.__init__(self, 'Amount')


    def __FixedData__(self, resultJson):
        result_subtotal, result_tip, result_tip_rate, result_total = self.__ParseData__(resultJson)

        logging.info(result_subtotal + u', ' + result_tip + u', ' + result_total + u' Fixed To ')

        result_tip_rate, tiprate_tip, tiprate_total = self.__FixedTipRate__(result_tip_rate)
        confidencelevel, result_subtotal, result_tip, result_total = self.__MildFixedAmountData__(result_subtotal, result_tip, result_total)
        if confidencelevel != ConfidenceLevel.Confident or confidencelevel != ConfidenceLevel.Fixed:
            confidencelevel, result_subtotal, result_tip, result_total = self.__FixedAmountData__(result_subtotal, result_tip, result_total, tiprate_tip, tiprate_total)

        logging.info(result_subtotal + u', ' + result_tip + u', ' + result_total)

        return confidencelevel, result_subtotal, result_tip, result_tip_rate, result_total


    def __FixedDataWithValidate__(self, resultJson, validateJson):
        result_subtotal, result_tip, result_tip_rate, result_total = self.__ParseData__(resultJson)
        validated_subtotal, validated_tip, validated_tip_rate, validated_total = self.__ParseData__(validateJson)

        if not self.__CheckData__(validated_subtotal) or not self.__CheckData__(validated_tip) or not self.__CheckData__(validated_total):
            logging.info(u'Validated Data Error')
            return

        if validated_subtotal != result_subtotal or validated_tip != result_tip or validated_total != result_total:
            self.__ErrorCount__ += 1
        else:
            logging.info(u'Validated Equal To Result')
            return

        logging.info(u'Validated Not Equal To Result')
        logging.info(result_subtotal + u', ' + result_tip + u', ' + result_total + u' Fixed To ')

        result_tip_rate, tiprate_tip, tiprate_total = self.__FixedTipRate__(result_tip_rate)
        confidencelevel, result_subtotal, result_tip, result_total = self.__MildFixedAmountData__(result_subtotal, result_tip, result_total)
        if confidencelevel != ConfidenceLevel.Confident or confidencelevel != ConfidenceLevel.Fixed:
            confidencelevel, result_subtotal, result_tip, result_total = self.__FixedAmountData__(result_subtotal, result_tip, result_total, tiprate_tip, tiprate_total)

        logging.info(result_subtotal + u', ' + result_tip + u', ' + result_total)

        if validated_subtotal == result_subtotal and validated_tip == result_tip and validated_total == result_total:
            self.__FixedCount__ += 1
            logging.info(u'Fixed Success!')
        else:
            logging.info(u'Validated ' + validated_subtotal + u', ' + validated_tip + u', ' + validated_total)
            logging.info(u'Fixed Falied!')


    def __ParseData__(self, jsondata):
        subtotal = ''
        tip = ''
        tip_rate = ''
        confidence = 0.0
        total = ''

        if jsondata == None or not isinstance(jsondata, dict) or jsondata[u'regions'] == None:
            return subtotal, tip, tip_rate, total

        regions = jsondata[u'regions']

        for region in regions:
            if region[u'cls'] == None or region[u'result'] == None:
                continue

            cls = region[u'cls']
            if cls == 3:
                subtotal = region[u'result'][0]
            elif cls == 5:
                tip = region[u'result'][0]
            elif cls == 6:
                total = region[u'result'][0]
            elif cls == 11 and float(region['confidence']) > confidence:
                tip_rate = region[u'result'][0]
                confidence = float(region['confidence'])

        return subtotal, tip, tip_rate, total


    def __CheckData__(self, data):
        if not self.__CheckFloat__(data):
            return False

        dotindex = data.find(u'.')
        if dotindex != -1 and (len(data) - dotindex) == 3:
            return True

        return False


    def __FixedTipRate__(self, tip_rate):
        if len(tip_rate) == 0:
            return u'', u'', u''

        tokens = []
        for token in re.split(u'%|\\$|,|:|-| ', tip_rate):
            if self.__CheckFloat__(token):
                tokens.append(token)

        if len(tokens) == 3:
            rate = int(tokens[0])
            if rate > 100:
                tokens[0] = str(rate % 100)

            return tokens[0] + u'% tip = $' + tokens[1] + u', Total = $' + tokens[2], tokens[1], tokens[2]

        return tip_rate, u'', u''


    def __MildFixedAmountData__(self, subtotal, tip, total):
        subtotal = self.__NormalizeData__(subtotal)
        tip = self.__NormalizeData__(tip)
        total = self.__NormalizeData__(total)

        if self.__CheckFloat__(subtotal) and self.__CheckFloat__(total):
            d_subtotal = Decimal(subtotal)
            d_total = Decimal(total)
            if self.__CheckFloat__(tip):
                d_tip = Decimal(tip)
                if d_total == (d_subtotal + d_tip):
                    return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(tip), self.__FormatData__(total)

                if self.__MissDot__(total):
                    d_temp = d_subtotal + d_tip
                    fixed = str(d_temp)
                    if StrUtil.Similarity(fixed, total) == 1 and len(fixed) - len(total) == 1:
                        if (d_temp == (d_subtotal + d_tip)):
                            return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(tip), self.__FormatData__(fixed)

                if self.__MissDot__(tip) or self.__MissInteger__(tip):
                    d_temp = d_total - d_subtotal
                    fixed = str(d_temp)
                    if StrUtil.Similarity(fixed, tip) == 1 and len(fixed) - len(tip) == 1:
                        if (d_total == (d_subtotal + d_temp)):
                            return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(fixed), self.__FormatData__(total)

            elif len(tip) == 0:
                if d_subtotal == d_total:
                    return ConfidenceLevel.Confident, self.__FormatData__(subtotal), tip, self.__FormatData__(total)
                elif self.__MissDot__(total) and d_total > d_subtotal:
                    totallen = len(total)
                    if totallen > 1:
                        pos = 1
                        while d_subtotal * 2 < d_total:
                            d_total = Decimal(total[0 : totallen - pos] + '.' + total[totallen - pos:])
                            pos += 1
                        if d_subtotal < d_total:
                            return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), tip, self.__FormatData__(str(d_total))

            else:
                d_temp = d_total - d_subtotal
                fixed = str(d_temp)
                if StrUtil.Similarity(fixed, tip) == 1:
                    return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(fixed), self.__FormatData__(total)

        return ConfidenceLevel.Bad, self.__FormatData__(subtotal), self.__FormatData__(tip), self.__FormatData__(total)


    def __FixedAmountData__(self, subtotal, tip, total, tiprate_tip, tiprate_total):
        confidencelevel, result_subtotal, result_tip, result_total = self.__FixedDataByTipRate__(subtotal, tip, total, tiprate_tip, tiprate_total)
        if confidencelevel == ConfidenceLevel.Confident or confidencelevel == ConfidenceLevel.Fixed:
            if len(tip) == 0:
                result_tip = u''
            if len(total) == 0:
                result_total = u''
            return confidencelevel, result_subtotal, result_tip, result_total

        if self.__CheckData__(subtotal):
            confidencelevel, result_subtotal, result_tip, result_total = self.__FixedDataBySubtotal__(subtotal, tip, total)
            if confidencelevel == ConfidenceLevel.Confident or confidencelevel == ConfidenceLevel.Fixed:
                if len(tip) == 0:
                    result_tip = u''
                if len(total) == 0:
                    result_total = u''
                return confidencelevel, result_subtotal, result_tip, result_total

        if self.__CheckData__(total):
            confidencelevel, result_subtotal, result_tip, result_total = self.__FixedDataByTotal__(subtotal, tip, total)
            if confidencelevel == ConfidenceLevel.Confident or confidencelevel == ConfidenceLevel.Fixed:
                if len(tip) == 0:
                    result_tip = u''
                if len(total) == 0:
                    result_total = u''
                return confidencelevel, result_subtotal, result_tip, result_total

        return ConfidenceLevel.Bad, subtotal, tip, total
    
    def __FixedDataByTipRate__(self, subtotal, tip, total, tiprate_tip, tiprate_total):
        if self.__CheckFloat__(tiprate_tip) and self.__CheckFloat__(tiprate_total):
            d_tiprate_total = Decimal(tiprate_total)
            d_tiprate_tip = Decimal(tiprate_tip)
            if self.__CheckFloat__(subtotal):
                d_subtotal = Decimal(subtotal)
                if (d_tiprate_total == (d_subtotal + d_tiprate_tip)):
                    if self.__CheckFloat__(total):
                        d_total = Decimal(total)
                        if d_tiprate_total == d_total:
                            return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(tiprate_tip), self.__FormatData__(total)
                        else:
                            if abs(d_tiprate_total - d_total) > min(d_tiprate_total, d_total):
                                d_total = d_tiprate_total
                            else:
                                d_total = min(d_tiprate_total, d_total)
                            if self.__CheckFloat__(tip):
                                d_tip = Decimal(tip)
                                if d_tip == d_tiprate_tip:
                                    return ConfidenceLevel.Confident, self.__FormatData__(subtotal), self.__FormatData__(tiprate_tip), self.__FormatData__(tiprate_total)
                                else:
                                    return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), self.__FormatData__(str(d_tiprate_total - d_subtotal)), self.__FormatData__(tiprate_total)
                            elif len(tip) == 0:
                                return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), tip, self.__FormatData__(str(d_total))
            else:
                d_subtotal = d_tiprate_total - d_tiprate_tip
                fixed = str(d_subtotal)
                if not self.__CheckFloat__(subtotal) or StrUtil.SimilarityRate(fixed, subtotal) > 0.7:
                    return ConfidenceLevel.Fixed, self.__FormatData__(fixed), self.__FormatData__(tiprate_tip), self.__FormatData__(tiprate_total)

        return ConfidenceLevel.Bad, u'', u'', u''


    def __FixedDataBySubtotal__(self, subtotal, tip, total):
        d_subtotal = Decimal(subtotal)
        tipTooLarge = False

        if self.__CheckFloat__(tip):
            d_tip = Decimal(tip)
            tipTooLarge = d_tip > (d_subtotal * Decimal(u'0.4'))
            if not tipTooLarge:
                d_total = d_subtotal + d_tip
                fixed = str(d_total)
                if d_total >= d_subtotal and (len(total) == 0 or not self.__CheckFloat__(total) or StrUtil.SimilarityRate(fixed, total) > 0.5):
                    if self.__CheckFloat__(total):
                        diff = d_total - Decimal(total)
                        if diff < Decimal('1.0') and diff > Decimal('0.0'):
                            fixed = total
                    return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), self.__FormatData__(tip), self.__FormatData__(fixed)

        if self.__CheckFloat__(total):
            d_total = Decimal(total)
            d_tip = d_total - d_subtotal
            fixed = str(d_tip)
            if d_tip >= Decimal(u'0.00') and (len(tip) == 0 or not self.__CheckFloat__(tip) or tipTooLarge or StrUtil.SimilarityRate(fixed, tip) > 0.5):
                return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), self.__FormatData__(fixed), self.__FormatData__(total)
            elif d_tip < Decimal(u'0.00'):
                fixed = subtotal;
                if StrUtil.SimilarityRate(fixed, total) > 0.6:
                    return ConfidenceLevel.Fixed, self.__FormatData__(subtotal), self.__FormatData__(u'0.00'), self.__FormatData__(fixed)

        return ConfidenceLevel.Bad, u'', u'', u''


    def __FixedDataByTotal__(self, subtotal, tip, total):
        d_total = Decimal(total)
        tipTooLarge = False

        if self.__CheckFloat__(tip):
            d_tip = Decimal(tip)
            tipTooLarge = d_tip > (d_total * Decimal(u'0.4'))
            if not tipTooLarge:
                d_subtotal = d_total - d_tip
                fixed = str(d_subtotal)
                if len(subtotal) == 0 or not self.__CheckFloat__(subtotal) or StrUtil.SimilarityRate(fixed, subtotal) > 0.5:
                    return ConfidenceLevel.Fixed, self.__FormatData__(fixed), self.__FormatData__(tip), self.__FormatData__(total)

        return ConfidenceLevel.Bad, u'', u'', u''


    def __NormalizeData__(self, amountdata):
        length = len(amountdata)
        if length == 0:
            return amountdata

        amountdata = re.sub(u'\\$| |%|\\\\', u'', amountdata)
        amountdata = ''.join([ch for ch in amountdata if (ch in self.__FloatPatterns__)])

        length = len(amountdata)
        dotindex = amountdata.find(u'.')
        if dotindex != -1:
            if (length - dotindex) > 2:
                return amountdata[0: dotindex + 3]

        return amountdata 


    def __FormatData__(self, amountdata):
        length = len(amountdata)
        if length == 0:
            return ''

        while len(amountdata) > 1 and amountdata[0] == u'0' and amountdata[1] == u'0':
            amountdata = amountdata[1:]

        dotindex = amountdata.find(u'.')
        if dotindex != -1:
            if (length - dotindex) > 2:
                return amountdata[0: dotindex + 3]
            elif (length - dotindex) == 2:
                return amountdata + u'0'
            elif (length - dotindex) == 1:
                return amountdata + u'00'
            return amountdata
        else:
            return amountdata + u'.00'


    def __MissDot__(self, amountdata):
        dotindex = amountdata.find(u'.')
        return dotindex == -1
    

    def __MissInteger__(self, amountdata):
        dotindex = amountdata.find(u'.')
        return dotindex == 0

